README.md
# Middleman-Sitemap-Xml-Generator


This extension is a forked and slightly extended version of [Middleman Search Engine Sitemap by Pete Nicholls](https://github.com/Aupajo/middleman-search_engine_sitemap).

Note that this extension has been forked from the release [tagged as v1.3.0](https://github.com/Aupajo/middleman-search_engine_sitemap/releases/tag/v1.3.0) to support Middleman-v3.

Please refer to [the original](https://github.com/Aupajo/middleman-search_engine_sitemap) if you want to get the latest release or a version for Middleman-v4.


### Why forked from v1.3.0?

Because I'm not be planning to use Middleman-v4 yet. 8-)


## Installation

Add the following line to the Gemfile of your Middleman project:

```ruby
gem 'middleman-sitemap-xml-generator'
```

Then as usual, run:

```sh
bundle install
```


## Usage

To activate and configure this extension, add the following configuration block to Middleman's config.rb:

```ruby
set :url_root, 'http://example.com'
activate :sitemap_xml_generator
```


### Options

| Option             | Default                | Description
| ---                | ---                    | ---
| default_changefreq | 'monthly'              | Default page priority
| default_priority   | 0.5                    | Default page priority
| sitemap_xml_path   | 'sitemap.xml'          | Path to the sitemap.xml
| exclude_attr       | 'hide_from_sitemap'    | Attribute in `Frontmatter` to exclude from the sitemap
| process_url        | nil                    | Callable object(responds to #call) to process a URL
| process_changefreq | nil                    | Callable object to determine a changefreq
| process_priority   | nil                    | Callable object to determine a priority
| exclude_if         | ->(resource) { false } | Callable object to exclude from the sitemap
| extra_ext          | []                     | Extensions allowed to be in the sitemap<br>(needed prefixed "dot")
| discard_ext        | []                     | Extensions removed from "loc" tag<br>(needed prefixed "dot")


## Determination of "changefreq" and "priority"

A value of `changefreq` and `priority` in the sitemap.xml is determined in the following priority.

1. **(Higher)** `Frontmatter` attributes in each file

    ```yaml
    ---
    changefreq: daily
    priority: 1.0
    ---
    ```

1. Processing in a `process_changefreq` and a `process_priority` option

    ```ruby
    activate :sitemap_xml_generator do |f|
      f.process_changefreq = ->(path, default) {
        path.end_with?('.foo') ? 'yearly' : default
      }
      f.process_priority = ->(path, default) {
        path.end_with?('.foo') ? 0.9 : default
      }
    end
    ```

    | Argument | Description
    | ---      | ---
    | path     | A path prefixed "/" from **Web** root. e.g. `/foo.html`
    | default  | The same value as `default_changefreq` or `default_priority`

1. **(Lower)** A value in a `default_changefreq` and a `default_priority` option


## Excluding pages


### Specify in Frontmatter

You can add a `hide_from_sitemap` attribute to `Frontmatter` in the page which is omitted from the sitemap.xml:

```yaml
---
hide_from_sitemap: true
---
```

If you would like to use a different `Frontmatter` attribute, you can specify it in a `exclude_attr` option:

```ruby
activate :sitemap_xml_generator do |f|
  f.exclude_attr = 'invisible'
end
```

Then you can omit a page the following way:

```yaml
---
invisible: true
---
```


### Use custom processor

You can also use a `exclude_if` option to exclude pages based on conditions you want:

```ruby
# Exclude all pages which have a date that's after today
activate :sitemap_xml_generator do |f|
  f.exclude_if = ->(resource) {
    resource.data.date && resource.data.date > Date.today
  }
end
```

| Argument | Description
| ---      | ---
| resource | A resource object (`Middleman::Sitemap::Resource`)


## Processing a URL in "loc" tag

You can use a `process_url` option to process a URL in "loc" tag in each page in the sitemap.xml:

```ruby
# This Proc removes a trailing slash from a URL in all "loc" tag.
activate :sitemap_xml_generator do |f|
  f.process_url = ->(url) { url.chomp('/') }
end
```

| Argument | Description
| ---      | ---
| url      | The URL of each page


## Adding file in addition to HTML

Only HTML file is added to the sitemap.xml with the default setting.

You can use a `extra_ext` option to add any file to the sitemap.xml:

```ruby
# This setting adds every file suffixed ".foo" to the sitemap.xml.
activate :sitemap_xml_generator do |f|
  f.extra_ext = ['.foo'] # prefixed dot is required.
end
```


## Excluding a content extension

You can use a `discard_ext` option to exclude a content extension from a address in "loc" tag in the sitemap.xml:

```ruby
# This setting removes suffixed ".foo" from "loc" tag in the sitemap.xml.
activate :sitemap_xml_generator do |f|
  f.discard_ext = ['.foo'] # prefixed dot is required.
end
```

Note that this process is executed before the process designated in a `process_url` option.


## Workaround for the template reference issue

This extension deals with that [issue](https://github.com/Aupajo/middleman-search_engine_sitemap/issues/5) by [this workaround](https://github.com/Aupajo/middleman-search_engine_sitemap/issues/2#issuecomment-46701165).


## Thanks

All ideas, logics, hints and prizes belong to the original author, Pete Nicholls.

And all mistakes belong to me.


## License

(c) 2016 AT-AT. MIT Licensed, see [LICENSE](LICENSE.md) for details.


CHANGELOG.md
## 1.0.1 (2020-06-12)

Update:

  - Change required version of "nokogiri".


spec/features/sitemap_feature_spec.rb
require 'spec_helper'
require 'nokogiri'
require 'pathname'

describe "Generate sitemap.xml", :feature do
  include XmlHelpers

  it "produces a valid sitemap" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :sitemap_xml_generator

      ignore '/ignored.html'
    end

    visit '/sitemap.xml'

    schema = Pathname(__dir__) + '../../sitemap.xsd'
    doc = Nokogiri::XML(last_response.body)
    expect(doc).to validate_against_schema(schema)

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/home.html',
      'priority'   => '0.5',
      'changefreq' => 'monthly'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/about.html',
      'priority'   => '0.2',
      'changefreq' => 'monthly'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/blog.html',
      'priority'   => '0.5',
      'changefreq' => 'daily'
    )

    expect(doc.to_s).not_to include('http://example.com/ignored.html')
    expect(doc.to_s).not_to include('http://example.com/ignored-in-frontmatter.html')
  end

  it "works with directory indexes" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :directory_indexes
      activate :sitemap_xml_generator

      ignore '/ignored.html'
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/home/',
      'priority'   => '0.5',
      'changefreq' => 'monthly'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/about/',
      'priority'   => '0.2',
      'changefreq' => 'monthly'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/blog/',
      'priority'   => '0.5',
      'changefreq' => 'daily'
    )

    expect(doc.to_s).not_to include('http://example.com/ignored/')
    expect(doc.to_s).not_to include('http://example.com/ignored-in-frontmatter/')
  end

  it "supports custom URL processing" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :directory_indexes
      activate :sitemap_xml_generator, process_url: ->(url) { url.upcase }
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc).to contain_node('url').with_children(
    'loc'        => 'HTTP://EXAMPLE.COM/HOME/',
    'priority'   => '0.5',
    'changefreq' => 'monthly'
    )
  end

  it "ignores resources based on exclude_if result" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :sitemap_xml_generator,
        exclude_if: ->(resource) { resource.data.ignored_by_proc }

      ignore '/ignored.html'
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc.to_s).not_to include('http://example.com/ignored-by-proc.html')
  end

  it "allows resources having a extension which is designated in the option" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :sitemap_xml_generator,
        extra_ext: ['.foo', 'bar']
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/allowed-extra-ext.foo',
      'priority'   => '0.2',
      'changefreq' => 'daily'
    )

    expect(doc.to_s).not_to include('http://example.com/allowed-extra-ext.bar')
  end

  it "removes designated extention from resources" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :sitemap_xml_generator,
        extra_ext: ['.foo', '.bar'],
        discard_ext: ['.html', '.foo', 'bar']
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/removed-ext',
      'priority'   => '0.2',
      'changefreq' => 'daily'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/removed-extra-ext',
      'priority'   => '0.2',
      'changefreq' => 'daily'
    )

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/removed-ext-remained.bar',
      'priority'   => '0.2',
      'changefreq' => 'daily'
    )
  end

  it "set values by procs" do
    run_site 'dummy' do
      set :url_root, 'http://example.com'
      activate :sitemap_xml_generator,
        extra_ext: ['.foo'],
        process_changefreq: ->(path, default){
          path.end_with?('.foo') ? 'yearly' : default
        },
        process_priority: ->(path, default){
          path.end_with?('.foo') ? 0.9 : default
        }
    end

    visit '/sitemap.xml'
    doc = Nokogiri::XML(last_response.body)

    expect(doc).to contain_node('url').with_children(
      'loc'        => 'http://example.com/values_by_proc.foo',
      'priority'   => '0.9',
      'changefreq' => 'yearly'
    )
  end

end


spec/support/middleman_server_helpers.rb
require 'middleman/rack'
require 'rack/test'

module MiddlemanServerHelpers
  include Rack::Test::Methods

  def app
    @app.call
  end

  def visit(path)
    get(path)
    raise last_response.errors if !last_response.errors.empty?
  end

  def find_on_page(string)
    expect(last_response.body).to include(string)
  end

  def run_site(path, &block)
    setup_environment(path)

    @app = lambda do
      instance = Middleman::Application.server.inst do
        # Require the extension after the server has booted, as would typically happen.
        require File.expand_path('../../../lib/middleman_extension', __FILE__)

        instance_exec(&block)
      end

      instance.class.to_rack_app
    end
  end

  private

  def setup_environment(path)
    ENV['MM_ROOT'] = File.expand_path("../../#{path}", __FILE__)
    ENV['TEST'] = "true"
  end
end


lib/middleman-sitemap-xml-generator/extension.rb
module Middleman
  module SitemapXmlGenerator
    class Extension < ::Middleman::Extension
      option :default_changefreq, 'monthly', 'Default page priority'
      option :default_priority, 0.5, 'Default page priority'
      option :sitemap_xml_path, 'sitemap.xml', 'Path to the sitemap.xml'
      option :exclude_attr, 'hide_from_sitemap', 'Attribute in Frontmatter to exclude from sitemap'
      option :process_url, nil, 'Callable object(responds to #call) to process a URL'
      option :process_changefreq, nil, 'Callable object to determine a changefreq'
      option :process_priority, nil, 'Callable object to determine a priority'
      option :exclude_if, ->(resource) { false }, 'Callable object to exclude from the sitemap'
      option :extra_ext, [], 'Extensions allowed to be in the sitemap(needed prefixed "dot")'
      option :discard_ext, [], 'Extensions removed from "loc" in the sitemap(needed prefixed "dot")'

      TEMPLATES_DIR = File.expand_path(File.join(File.dirname(__FILE__), 'templates'))


      def initialize(app, options_hash={}, &block)
        super

        require 'builder'
      end


      public


      #
      # Hooks
      #

      def after_configuration
        register_extension_templates
        prepare_extra_ext_opt
        prepare_discard_ext_opt
        watch_vendor_directory
      end

      def manipulate_resource_list(resources)
        resources << sitemap_resource
      end


      #
      # Helpers
      #

      def resource_in_sitemap?(resource)
        is_page?(resource) && not_excluded?(resource)
      end

      def process_changefreq(path)
        default = options.default_changefreq
        options.process_changefreq ? options.process_changefreq.call(path, default) : default
      end

      def process_priority(path)
        default = options.default_priority
        options.process_priority ? options.process_priority.call(path, default) : default
      end

      def process_url(path)
        path = process_extension path
        options.process_url ? options.process_url.call(path) : path
      end

      helpers do
        def resources_for_sitemap
          sitemap.resources.select do |resource|
            extensions[:sitemap_xml_generator].resource_in_sitemap?(resource)
          end
        end
      end


      #
      # Internals
      #

      private

      def allowed_ext?(resource)
        options.extra_ext.include? File.extname(resource.path)
      end

      def filter_valid_extension(value)
        [value].flatten.select { |ext| ext.match /^\.[^.]+/ }
      end

      def generate_processor(method_name)
        self.class.instance_method(method_name.to_sym).bind(self)
      end

      def is_page?(resource)
        resource.path.end_with?(page_ext) || allowed_ext?(resource)
      end

      def not_excluded?(resource)
        !resource.ignored? \
          && !resource.data[options.exclude_attr] && !options.exclude_if.call(resource)
      end

      def page_ext
        File.extname(app.index_file)
      end

      def prepare_extra_ext_opt
        options.extra_ext = filter_valid_extension options.extra_ext
      end

      def prepare_discard_ext_opt
        options.discard_ext = filter_valid_extension options.discard_ext
      end

      def process_extension(path)
        target_ext = File.extname path
        if options.discard_ext.include?(target_ext)
          path = File.join(File.dirname(path), File.basename(path, target_ext))
        end
        path
      end

      def register_extension_templates
        # We call reload_path to register the templates directory with Middleman.
        # The path given to app.files must be relative to the Middleman site's root.
        templates_dir_relative_from_root = \
          Pathname(TEMPLATES_DIR).relative_path_from(Pathname(app.root))
        app.files.reload_path(templates_dir_relative_from_root)
      end

      def sitemap_locals
        {
          url_proc: generate_processor(:process_url),
          changefreq_proc: generate_processor(:process_changefreq),
          priority_proc: generate_processor(:process_priority)
        }
      end

      def sitemap_resource
        source_file = template('sitemap.xml.builder')

        ::Middleman::Sitemap::Resource.new(app.sitemap, sitemap_xml_path, source_file).tap do |res|
          res.add_metadata(options: { layout: false }, locals: sitemap_locals)
        end
      end

      def sitemap_xml_path
        options.sitemap_xml_path
      end

      def template(path)
        full_path = File.join(TEMPLATES_DIR, path)
        raise "Template #{full_path} not found" if !File.exist?(full_path)
        full_path
      end

      # Workaround for the template reference issue. See below:
      #   https://github.com/Aupajo/middleman-search_engine_sitemap/issues/5
      #   https://github.com/Aupajo/middleman-search_engine_sitemap/issues/2#issuecomment-46701165
      def watch_vendor_directory
        app.config[:file_watcher_ignore] = (app.config[:file_watcher_ignore].reject do |regexp|
          regexp == /^vendor(\/|$)/
        end)
      end

    end
  end
end
